lib/ibm/cloud/sdk/vpc/instances.rb

```diff
@@ -38,21 +38,111 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
 
         # Work with a single instance.
         class Instance < VPCInstance
+          TRANSITIONAL_STATES = %(pausing pending restarting resuming starting stopping)
+          ERROR_STATE = 'failed'
+          RUNNING_STATE = 'running'
+          STOPPED_STATES = %('stopped', 'paused')
+
+          # The id of this VM.
+          def id
+            @data[:id]
+          end
+
+          # The name of this VM.
+          def name
+            @data[:name]
+          end
+
+          # The status of the virtual server instance. Possible values: [failed,paused,pausing,pending,restarting,resuming,running,starting,stopped,stopping]
+          def status
+            @data[:status]
+          end
+
+          # Whether the state of the VM is in failed state.
+          # @return [Boolean]
+          def failed?
+            status == ERROR_STATE
+          end
+
+          # Whether the state of the VM is in the started state.
+          # @return [Boolean]
+          def started?
+            status == RUNNING_STATE
+          end
+
+          # Whether the state of the VM is in a stopped or paused state.
+          # @return [Boolean]
+          def stopped?
+            STOPPED_STATES.include?(status)
+          end
+
+          # Whether the state of the VM is in a transitional state.
+          # @return [Boolean]
+          def transtional?
+            TRANSITIONAL_STATES.include?(status)
+          end
+
+          # Interact with instance actions.
+          # @return [INSTANCE::Actions]
           def actions
             INSTANCE::Actions.new(self)
           end
 
+          # Interact with instance network interfaces.
+          # @return [INSTANCE::NetworkInterfaces]
           def network_interfaces
             INSTANCE::NetworkInterfaces.new(self)
           end
 
+          # Interact with instance volume attachements.
+          # @return [INSTANCE::VolumeAttachments]
           def volume_attachments
             INSTANCE::VolumeAttachments.new(self)
           end
 
+          # Return the data used for initializing this VM.
           def initialization
             adhoc(method: 'get', path: 'initialization').json
           end
+
+          # Wait for the VM instance to be in a stable state.
+          # @param started [Boolean] When true wait for started, when false wait for stopped.
+          # @param sleep_time [Integer] The time to sleep between refreshes.
+          # @param timeout [Integer] The number of seconds before raising an error.
+          # @return [Boolean] The return of the status check.
+          # @raise [RuntimeError] Instance goes into failed state.
+          # @raise [RuntimeError] Timeout has been reached.
+          def wait_for(started: true, sleep_time: 5, timeout: 600)
+            refresh
+            # When in transitional state wait for it to be stable.
+            until test_state(started)
```

Comment:
We could generalize this a bit if we accepted a block from the caller so: `wait_for(&:stopped?)` which makes more intuitive sense than `wait_for(started:false)`.

This way we can support any target state (e.g. not sure if VPC supports suspending or a suspended state) without adding more boolean arguments.  Might be a moot point if the only options are running or stopped.

Reply:
Hah, yeah, I was thinking of adding a block over the weekend. I remembered that the PureApp version took a block which was useful. It would be nice to have the option to modify the logic without updating the method. I can look at it later. Using the boolean was to avoid typos and reduce complexity when setting the state to look for.